### ISSUE TITLE
Fine-tuning Qwen3-TTS-12Hz-1.7B-Base for Bengali: Foreign/Hindi Accent Bias and Lack of Naturalness

### ISSUE THREAD

**Issue by kawshikbuet17** [QwenLM/Qwen3-TTS#323]
### Description

# Fine-tuning Qwen3-TTS-12Hz-1.7B-Base for Bengali: Foreign/Hindi Accent Bias and Lack of Naturalness

## Problem

I fine-tuned `Qwen3-TTS-12Hz-1.7B-Base` on a Bengali / Bangla dataset of around 7 hours from a single speaker. I am planning to train on 150+ hours of multi-speaker Bengali data later.

For inference, I used the `custom_voice` type model functions. The fine-tuning process itself runs successfully after applying a few patches, and the trained model is able to generate Bengali speech.

However, the generated speech does not sound like native Bengali. It sounds more like Bengali spoken with a Foreign/Hindi accent or Hindi-influenced prosody. The pronunciation is somewhat Bengali-like, but the accent, rhythm, intonation, and speaking style are noticeably Hindi-influenced and not natural for Bengali.

## Dataset

- Language: Bengali / Bangla
- Dataset size: around 7 hours
- Number of samples: 8k+
- Speaker setup: single Bengali female speaker
- Text format: Bengali letters and punctuation
- Audio code fields: prepared using the Qwen3 TTS tokenizer

We also tried IPA / romanized phoneme input. Interestingly, the Hindi accent issue was reduced with phoneme-style input, but the generated speech became more robotic and less natural compared to using Bengali script. Because of that, we are keeping the phoneme approach aside for now and may revisit it later with a larger dataset.

## Training Setup

- Base model: `Qwen3-TTS-12Hz-1.7B-Base`
- Speaker name: `bengali_female`
- GPU: RTX 3090
- Batch size: `2`
- Learning rate: `2e-6`
- Higher learning rate used for:
  - text embeddings
  - codec embeddings
  - some upper / adaptation layers
- Goal of higher LR on selected layers: stronger Bengali accent, tone, and prosody adaptation
- Epochs: `10`
- Warmup: `3%`
- Scheduler: Sequential Linear
- Validation ratio: `0.01`
- Inference language: `Auto`

## Observed Behavior

- Training loss decreases normally at the beginning.
- After around 2–3 epochs, the loss becomes mostly stuck.
- After that point, there is not much improvement in speech quality.
- In some cases, quality starts to degrade with further training.
- The model generates Bengali speech, but there are still some pronunciation issues.
- The main problem is that the accent does not sound like native Bengali.
- Prosody, rhythm, and speaking style sound closer to Hindi than Bengali.
- The generated speech lacks naturalness compared to native Bengali speech.

## Questions

1. **Does `Qwen3-TTS-12Hz-1.7B-Base` have a strong prior bias toward Hindi or other Indic-language prosody?**  
   As far as I know, the base TTS model may not have seen Hindi speech directly, but the underlying LLM may have Hindi text bias. Since Hindi and Bengali scripts / phonetics have some overlap, could the model be mapping Bengali text through Hindi-like phonetic or prosodic patterns?

2. **What is the recommended way to achieve stronger Bengali accent adaptation and more natural Bengali prosody?**  
   Are there specific layers, learning-rate strategies, prompt formats, speaker conditioning methods, or data-preparation changes that help with native accent adaptation for a new language?

3. **Should we add an explicit Bengali language tag during fine-tuning and inference?**  
   For example, would adding something like `<lang:bn>` or another Bengali language identifier help the model separate Bengali from Hindi / other Indic-language priors?

4. **Are there recommended fine-tuning parameters or guideline for unsupported or low-resource languages?**  

5. **Is there a recommended Bengali G2P / phoneme / romanization pipeline for Qwen3 TTS fine-tuning?**  
   Raw Bengali script gives more natural speech but Hindi-influenced accent. IPA / romanized phoneme input reduces the Hindi accent but sounds more robotic. Is there a better Bengali phoneme representation, G2P pipeline, or hybrid text + phoneme approach recommended for new-language fine-tuning?

## Additional Notes

The main goal is to fine-tune Qwen3 TTS for native-sounding Bengali speech with natural Bengali accent, rhythm, and intonation. Any guidance on dataset preparation, Bengali text normalization, phoneme design, language tagging, or fine-tuning strategy would be very helpful.

Thanks in advance!

**Comment by arpit763** [comment]
hi, 
would it be possible you to share the script, i am also planning to finetune on my data for a quick experiment.

**Comment by arpit763** [comment]
was it the official code, as i don't see any lora config there?

**Comment by kawshikbuet17** [comment]
@arpit763 , yes I have used the repository provided finetune scripts (https://github.com/QwenLM/Qwen3-TTS/tree/main/finetuning) and modified a little bit for my use case.

**Comment by mozi1924** [comment]
@kawshikbuet17 Sorry for the delayed response, and thank you for providing such a detailed description of the issue.

Apologies for the "glue code" I wrote; I inferred the following results based on the Qwen3-TTS source code. :(

### Root Cause Analysis

The language control mechanism in Qwen3-TTS relies heavily on `codec_language_id`—which is a discrete token embedded in the **codec channel**, not the text channel.

When you set `language="Auto"`, the model receives a codec prefill sequence that looks like this:

```
[nothink_id, think_bos_id, think_eos_id]   ← No language tag included

```

However, when you explicitly specify a language (e.g., `language="Chinese"`), the sequence changes to:

```
[think_id, think_bos_id, language_id, think_eos_id]   ← Language tag included

```

**The current official fine-tuning script (`dataset.py`) hardcodes the `nothink` mode**, completely ignoring the `language` field in the dataset. This means that no matter what language data you use for fine-tuning, the model never explicitly learns "this is Bengali." Instead, it is forced to guess based purely on the statistical patterns of the text/audio. Because Bengali and Hindi share visual/phonetic overlaps, the model easily falls into a Hindi accent bias.

---

### Solution: Adding an Explicit Bengali Tag for Fine-Tuning

Here is the complete step-by-step guide to fixing this:

#### Step 1: Register Bengali in `config.json`

In your base model's `config.json`, locate `talker_config.codec_language_id` and add a mapping for Bengali. Choose an unused codec token ID (we recommend an unassigned position in the 2048-4096 range):

```json
"talker_config": {
  "codec_language_id": {
    "chinese": ..., 
    "english": ...,
    "bn": 4206
  },
  ...
}

```

#### Step 2: Modify `collate_fn` in `dataset.py`

The current hardcoded codec prefill (lines 175-181 in `dataset.py`) looks like this:

```python
input_ids[i, 3:8, 1] = torch.tensor([
    self.config.talker_config.codec_nothink_id,
    self.config.talker_config.codec_think_bos_id,
    self.config.talker_config.codec_think_eos_id,
    0,
    self.config.talker_config.codec_pad_id
])

```

This needs to be dynamically constructed based on the `language` field.

> **Note:** When a language tag is present, the prefill length increases from 5 to 6 (due to the addition of `language_id`), meaning all subsequent offsets will need to be adjusted.

First, update the dataset's `__getitem__` method to return the `language` field, which it currently reads but drops:

```python
# Inside __getitem__
return {
    "language": language,  # Add this
    ...
}

```

Then, handle this in `collate_fn`:

```python
lang = data.get('language', 'Auto')
if lang.lower() == 'auto':
    codec_prefill = [
        self.config.talker_config.codec_nothink_id,
        self.config.talker_config.codec_think_bos_id,
        self.config.talker_config.codec_think_eos_id,
        0,  # Placeholder for speaker embedding
        self.config.talker_config.codec_pad_id
    ]
    prefill_len = 5
else:
    lang_id = self.config.talker_config.codec_language_id[lang.lower()]
    codec_prefill = [
        self.config.talker_config.codec_think_id,
        self.config.talker_config.codec_think_bos_id,
        lang_id,
        self.config.talker_config.codec_think_eos_id,
        self.config.talker_config.codec_pad_id
    ]
    prefill_len = 5  # Speaker placeholder is not needed here when language is specified
    # Note: The speaker embedding will need to be injected elsewhere.

```

**Crucial Challenge:** When the language tag is present, the prefill sequence structure changes, and the injection point for the speaker embedding also shifts (refer to the `codec_input_embedding` concatenation logic in the `generate()` method). You will need to carefully align these.

#### Step 3: Use the Language Tag during Inference

Once fine-tuning is complete, you must explicitly pass `language="bn"` (or whatever name you registered in the config) during inference:

```python
wavs, sr = tts.generate_custom_voice(
    text="আমি বাংলায় কথা বলি。",
    language="bn",        # ← Explicitly specify Bengali
    speaker="bengali_female",
)

```

---

### Addressing Your Specific Questions

1. **Does Qwen3-TTS-Base have a prior bias toward Hindi?** Yes. The underlying LLM (Qwen3) has significantly more Hindi text than Bengali in its pre-training data, and both scripts (Devanagari vs. Bengali) share some visual similarities in their Unicode blocks. Without an explicit language tag, the model naturally defaults to the more dominant Hindi phonetic patterns found in its training data when processing Bengali text.
2. **How can we achieve better Bengali accent adaptation?** * **Most Critical:** Add the explicit language tag as described above.
* **Scale Data:** Scaling up your dataset (your planned 150+ hours) is absolutely the right direction.
* **Hyperparameters:** Using a higher learning rate for `codec_embedding` and `text_embedding` is a reasonable strategy to force the model to adapt quickly.


3. **Should we add an explicit Bengali tag?** **Yes, absolutely.** This is the core fix. The language tag signals the model that "this is a distinct language," preventing it from relying on character-level statistical guessing.
4. **Hyperparameter recommendations for low-resource languages?** * 7 hours of single-speaker data is quite small; we recommend at least 20+ hours.
* The loss plateauing after 2-3 epochs suggests the data volume is insufficient for the model to generalize further. Try lowering the learning rate or using a longer warmup period.
* Your upcoming 150-hour multi-speaker dataset will significantly improve performance.


5. **Bengali G2P / Phoneme pipeline recommendations?** There is currently no officially recommended Bengali G2P pipeline. Your observation that "phonemes reduce the accent but sound robotic" occurs because phonemic representation loses semantic/contextual information that the LLM leverages from raw text. **Our recommended approach is to stick with raw Bengali text combined with the explicit language tag fine-tuning.** This teaches the model to directly associate Bengali text with Bengali speech patterns.

---

### Additional Modifications for `sft_12hz.py`

Aside from `dataset.py`, you will also need to adjust the embedding construction logic in `sft_12hz.py` (currently, the speaker embedding position is hardcoded at `[:, 6, :]` around lines 92-96). This needs to be dynamic to accommodate the different sequence lengths caused by adding/omitting the language tag.

**Comment by ravk1234** [comment]
@kawshikbuet17 Did you try this method suggested by @mozi1924 ? I am also finetuning the Qwen 3 TTS in Indian languages wanted to know if the audio was good this time or not.

**Comment by kawshikbuet17** [comment]
> [@kawshikbuet17](https://github.com/kawshikbuet17) Did you try this method suggested by [@mozi1924](https://github.com/mozi1924) ? I am also finetuning the Qwen 3 TTS in Indian languages wanted to know if the audio was good this time or not.

@ravk1234 Not yet. If you tried for indian languages, please let me know the update.

**Comment by MinhazurSarker** [comment]
@kawshikbuet17 Brother, I used your [kawshikbuet17/bengali-cosyvoice3-tts](https://huggingface.co/kawshikbuet17/bengali-cosyvoice3-tts) and it is Good .. Loved it.. also made a set of ComfyUI custom nodes.. it was good but worked only for small texts.. and it didn't support longer reference audio & transcripts. but it was good though..

I am very optimistic about your work on Qwen3 TTS.

Love from Your Bangladeshi Brother <3

(I'm not a AI engineer nor a Python programmer. My works are only focused on building business softwares using Node js and sometimes mix of others)
